REGRESSION(258561@main): :user-valid/:user-invalid stopped working

shvaikalesh opened the pull request:
#### bc8d5b0f582f1f25dd48a6d84f51aeb29dd1e20b
<pre>
REGRESSION(258561@main): :user-valid/:user-invalid stopped working
<a href="https://bugs.webkit.org/show_bug.cgi?id=254571">https://bugs.webkit.org/show_bug.cgi?id=254571</a>
&lt;rdar://problem/107300848&gt;

Reviewed by Tim Nguyen.

This change adds missing virtual method overrides to fix :user-valid / :user-invalid
pseudoclasses, similar to those we have for :valid / :invalid.
They are necessary because ValidatedFormListedElement doesn&apos;t inherit from Element.

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-valid.html:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/selectors/user-invalid-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/selectors/user-valid-expected.txt:
* Source/WebCore/html/HTMLFormControlElement.h:

Co-authored-by: Tim Nguyen &lt;ntim@apple.com&gt;

Canonical link: <a href="https://commits.webkit.org/262512@main">https://commits.webkit.org/262512@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1371afd1bb85309a5f5dcfc25828176bfec6ef65

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1764 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1795 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1853 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2692 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1891 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1746 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1866 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1862 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1654 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1782 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1597 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/1582 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2531 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1590 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1568 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1496 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1624 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1611 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2695 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1654 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1471 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1577 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1575 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1716 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/183 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-invalid.html:
```diff
@@ -12,6 +32,31 @@
+promise_test(async () => {
+  const input = document.querySelector("#user-invalid-test");
+  assert_false(input.matches(":user-valid"), "Initially does not match :user-valid");
+  assert_false(input.matches(":user-invalid"), "Initially does not match :user-invalid");
```
The previous subtest could probably be combined with this one (moving the input validity check) and the other `<input>` removed.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-valid.html:
```diff
@@ -1,14 +1,59 @@
+</style>
+
 <input>
+<input id="user-valid-test" type="email">
```
Can we change the IDs for both inputs? (`initially-valid` & `initially-invalid`) The first one initially matches `input.validity.valid` whereas the second one doesn't.

nt1m reviewed (approved):
Thanks!

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-invalid.html:
```diff
@@ -12,6 +32,31 @@
+  assert_true(input.matches(":user-invalid"), "Typed an invalid email, :user-invalid now matches");
+  assert_false(input.matches(":user-valid"), "Typed an invalid email, :user-valid does not match");
+
+
```
nit: extra new line here

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-valid.html:
```diff
@@ -1,14 +1,59 @@
+  assert_true(input.matches(":user-valid"), "Typed a valid email, :user-valid now matches");
+  assert_false(input.matches(":user-invalid"), "Typed a valid email, :user-invalid does not match");
+
+
```
nit: extra new line here

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-invalid.html:
```diff
@@ -12,6 +32,31 @@
+
+  input.value = "";
+  assert_true(input.matches(":user-valid"), "Put a valid email, :user-valid now matches");
+  assert_false(input.matches(":user-invalid"), "Put an valid email, :user-invalid no longer matches");
```
We just cleared the input here, not enter a valid email.

shvaikalesh replied:
Added `input.required = true`.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-invalid.html:
```diff
@@ -1,16 +1,58 @@
+  input.blur();
+
+  assert_true(input.matches(":user-valid"), "Put a valid email, :user-valid now matches");
+  assert_false(input.matches(":user-invalid"), "Put an valid email, :user-invalid no longer matches");
```
This looks suspicious, doesn't the value in the input end up being `not an email test@example.com` (which is invalid?).

shvaikalesh replied:
Oh thanks Tim, that is a great catch and it's probably the reason WPE EWS is failing. Added `input.value = "";`, let's see if it helps.

**EDIT**: it does!

Commit:
REGRESSION(258561@main): :user-valid/:user-invalid stopped working
https://bugs.webkit.org/show_bug.cgi?id=254571
<rdar://problem/107300848>

Reviewed by Tim Nguyen.

This change adds missing virtual method overrides to fix :user-valid / :user-invalid
pseudoclasses, similar to those we have for :valid / :invalid.
They are necessary because ValidatedFormListedElement doesn't inherit from Element.

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/user-valid.html:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/selectors/user-invalid-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/selectors/user-valid-expected.txt:
* Source/WebCore/html/HTMLFormControlElement.h:

Co-authored-by: Tim Nguyen <ntim@apple.com>

Canonical link: https://commits.webkit.org/262512@main
